rainier-cats/src/main/scala/com/stripe/rainier/cats/package.scala

```diff
@@ -1,19 +1,54 @@
 package com.stripe.rainier.cats
 
 import com.stripe.rainier.compute.Real
-import com.stripe.rainier.core.Generator
-import com.stripe.rainier.core.RandomVariable
+import com.stripe.rainier.core.{Categorical, Generator, RandomVariable}
 import com.stripe.rainier.sampler.RNG
 import _root_.cats.Monad
 
 import scala.annotation.tailrec
 
 object `package` {
+  implicit val rainierMonadCategorical: Monad[Categorical] =
+    new MonadCategorical
   implicit val rainierMonadGenerator: Monad[Generator] = new MonadGenerator
   implicit val rainierMonadRandomVariable: Monad[RandomVariable] =
     new MonadRandomVariable
 }
 
+private[cats] class MonadCategorical extends Monad[Categorical] {
+  def pure[A](x: A): Categorical[A] = Categorical(Map(x -> Real.one))
+
+  override def map[A, B](fa: Categorical[A])(f: A => B): Categorical[B] =
+    fa.map(f)
+
+  override def product[A, B](
+      fa: Categorical[A],
+      fb: Categorical[B]
+  ): Categorical[(A, B)] = fa.zip(fb)
+
+  override def flatMap[A, B](fa: Categorical[A])(
+      f: A => Categorical[B]): Categorical[B] =
+    fa.flatMap(f)
+
+  def tailRecM[A, B](a: A)(
+      f: A => Categorical[Either[A, B]]): Categorical[B] = {
+
+    @tailrec
+    def run(acc: Map[B, Real],
+            pairs: List[(Either[A, B], Real)]): Map[B, Real] = {
+      pairs match {
+        case Nil => acc
+        case (Left(a), r) :: xs =>
+          run(acc, xs ++ f(a).pmf.mapValues(_ * r).toList)
```

Comment:
1. I'd use `xs ::: f(a).pmf.mapValues(_ * r).toList` so we don't use the `Iterable` method and we use the List specific version, may not matter.
2. more concerning, this can be `O(N^2)` since it is appending on the right. One way to solve this would be to use a `Queue[(Either[A, B], Real)]`. instead of List.

Reply:
I like the suggestion - I've switched the implementation to use a `Queue`. Thanks!